api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java

```diff
@@ -10,28 +10,181 @@
+import org.hibernate.criterion.Criterion;
+import org.hl7.fhir.exceptions.FHIRException;
+import org.hl7.fhir.r4.model.AllergyIntolerance;
+import org.openmrs.AllergenType;
 import org.openmrs.Allergy;
+import org.openmrs.Concept;
+import org.openmrs.module.fhir2.FhirConstants;
+import org.openmrs.module.fhir2.api.FhirConceptService;
+import org.openmrs.module.fhir2.api.FhirGlobalPropertyService;
 import org.openmrs.module.fhir2.api.dao.FhirAllergyIntoleranceDao;
 import org.springframework.stereotype.Component;
 
 @Component
 @Setter(AccessLevel.PACKAGE)
-public class FhirAllergyIntoleranceDaoImpl implements FhirAllergyIntoleranceDao {
+public class FhirAllergyIntoleranceDaoImpl extends BaseDaoImpl implements FhirAllergyIntoleranceDao {
 	
 	@Inject
 	@Named("sessionFactory")
 	private SessionFactory sessionFactory;
 	
+	@Inject
+	private FhirConceptService conceptService;
+	
+	@Inject
+	private FhirGlobalPropertyService globalPropertyService;
+	
 	@Override
 	public Allergy getAllergyIntoleranceByUuid(String uuid) {
 		return (Allergy) sessionFactory.getCurrentSession().createCriteria(Allergy.class).add(eq("uuid", uuid))
 		        .uniqueResult();
 	}
+	
+	@Override
+	public Collection<Allergy> searchForAllergies(ReferenceParam patientReference, StringOrListParam category,
+	        TokenOrListParam allergen, TokenOrListParam severity, TokenOrListParam manifestationCode,
+	        StringOrListParam clinicalStatus) {
+		Criteria criteria = sessionFactory.getCurrentSession().createCriteria(Allergy.class);
+		handlePatientReference(criteria, patientReference, "patient");
+		handleAllergenCategory("allergen.allergenType", category);
+		handleAllergen(criteria, allergen);
+		handleSeverity(criteria, severity).ifPresent(criteria::add);
+		handleManifestation(criteria, manifestationCode);
+		handleBoolean("voided", createClinicalStatusToken(clinicalStatus));
+		
+		return criteria.list();
+	}
+	
+	private void handleManifestation(Criteria criteria, TokenOrListParam manifestation) {
+		if (manifestation == null) {
+			return;
+		}
+		criteria.createAlias("reactions", "r");
+		criteria.createAlias("r.reaction", "c");
+		
+		handleOrListParamBySystem(manifestation, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(in("c.uuid", tokensToList(tokens)));
+			} else {
+				if (!containsAlias(criteria, "rs")) {
+					criteria.createAlias("c.uuid", "rs");
+				}
+				
+				return Optional.of(or(eq("rs.name", system), in("c.uuid", tokensToList(tokens))));
+			}
+		}).ifPresent(criteria::add);
+	}
+	
+	private void handleAllergen(Criteria criteria, TokenOrListParam allergen) {
+		if (allergen == null) {
+			return;
+		}
+		
+		criteria.createAlias("allergen.codedAllergen", "c");
+		
+		handleOrListParamBySystem(allergen, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(in("c.uuid", tokensToList(tokens)));
+			} else {
+				if (!containsAlias(criteria, "cs")) {
+					criteria.createAlias("c.uuid", "cs");
```

Comment:
This seems to be creating a join on the `codedAllergen.uuid` field, which is not going to work.